src/Voron/Data/Tables/Table.cs

```diff
@@ -1684,30 +1684,153 @@ public bool SeekOnePrimaryKeyWithPrefix(Slice prefix, Slice value, out TableValu
+            private int _idx;
+
+            public PagePrefetcherIterator(ByteString locations, int documents)
+            {
+                Debug.Assert(locations.Length >= documents);
+
+                _locations = locations;
+                _documents = documents;
+                _idx = -1;
+            }
+
+            long IEnumerator<long>.Current => ((long*)_locations.Ptr)[_idx] / Constants.Storage.PageSize;
+
+            object IEnumerator.Current => ((long*)_locations.Ptr)[_idx] / Constants.Storage.PageSize;
+
+            void IDisposable.Dispose() { }
+
+            bool IEnumerator.MoveNext()
+            {
+                _idx++;
+                return _idx < _documents;
+            }
+
+            void IEnumerator.Reset()
+            {
+                _idx = -1;
+            }
+        }
+
+        public IEnumerable<(long Key, TableValueHolder TableValueHolder)> IterateForDictionaryTraining(FixedSizeKeyIndexDef index, long skip = 1, long seek = 0)
         {
             if (skip < 1)
                 throw new ArgumentOutOfRangeException(nameof(skip), "The skip must be positive and non zero.");
 
             var fst = GetFixedSizeTree(index);
 
-            long count = -1;
-            using var it = fst.Iterate();
+            // Dictionary training can be extremely IO intensive, therefore as illustrated in RavenDB-21106 we need to have
+            // an heuristic that allows us to get good dictionaries but at the same time minimizing the IO usage specially
+            // in IO deprived setups.
+            // https://issues.hibernatingrhinos.com/issue/RavenDB-21106
+
+            // If we page fault on every page (specially in cold scenarios) we would be wasting a lot. Therefore we should
+            // iterate with prefetching enabled.
+            using var it = fst.Iterate(prefetch: true);
             if (it.Seek(seek) == false)
                 yield break;
 
-            var result = new TableValueHolder();
-            while (it.MoveNext())
+            // Considering that in big collections the max amount of documents has been defined to be in the order of 100K 
+            // If we have to big skips we should get as many documents as possible to obtain as high locality as data allows.
+            int ChunkSize = 1024 * 16;
+
+            int ReadChunked(FixedSizeTree<long>.IFixedSizeIterator it, ByteString chunk, out bool hasMore)
             {
-                count++;
-                if (count % skip == 0)
+                // We will fill the buffer of the chunk expecting to have many documents in the chunk that end up 
+                // being stored in segments that are close to each other.
+                var chunkSpan = new Span<long>(chunk.Ptr, ChunkSize);
+                var keySpan = new Span<long>(chunk.Ptr + ChunkSize * sizeof(long), ChunkSize);
+                Debug.Assert(chunkSpan.Length == ChunkSize);
+                Debug.Assert(keySpan.Length == ChunkSize);
```

Comment:
You are checking the length of the _spans_, but you need to check the size of _chunk_.